Move log timestamp to Foundation Date

JGiola opened the pull request:
# Summary

This pull request is aimed to the issue #129 for using Foundation.Date for log timestamps
# Design decisions

I've add a private time formatter shared instance for keeping the initialisation cost low, and set the format template to print hours, minutes, seconds and milliseconds.
I don't know if you prefer additional information on the date or the time is enough.

JGiola committed:
Move log timestamp to Foundation Date

Move the currentTime function in Log.swift from timeval to Foundation
Date with custom string formatter

dreymonde commented:
Nice job, thanks! However, I think it is more appropriate to set date format like this:

``` swift
formatter.dateFormat = DateFormatter.dateFormat(fromTemplate: "HH:mm:ss.SSS", options: 0, locale: .current)
```

This way, the user's locale will be respected. For example, for me it prints time like this:

> 13:52:19,075

(with comma after 19, not the dot), which is appropriate in my region (Ukraine).

dreymonde reviewed (commented):
Also, maybe it's better to put `dateFormatter` inside `Logger` as `fileprivate` property.

JGiola commented:
Oh, ok, I've done that at first but then I've gone to do a fixed format.

The dateFormatter property is outside Logger for having it shared across all Logger instances and not create a new one every time a Logger object is created.

I will push the changes when I will be back home.

JGiola committed:
Change dateFormat template

Move from a fixed template to a localised one through the use of
dateFormat(fromTemplate:options:locale:) static method on DateFormatter

JGiola commented:
While I'm at it, you think I have to change the `HH` part of the template with `jj` to adapt dinamically to 12/24 hour clock?
Or at least is what I can understand from this [documentation](http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns) even if I don't know if you have to add the `a` to add AM/PM...

JGiola committed:
Move logTimeFormatter property

Move the property inside the Logger class declaration

Danappelxx commented on Modules/Axis/Sources/Axis/Log/Log.swift:
```diff
@@ -51,6 +47,13 @@ public final class Logger {
+
+    fileprivate static let logTimeFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        let format = DateFormatter.dateFormat(fromTemplate: "HH:mm:ss.SSS", options: 0, locale: .current)
```
Could we also include the day/month/year and locale here? I think that would be useful. Ideal date IMO: `10/06/16 12:42:10.123 GMT-8`.

Note - I'm not sure if `DateFormatter` supports figuring out whether the month should go before or in-between the day and year (in the US, we have the day in the middle).

I also think we should use the 24-hour system regardless of the locale, so the code you have now (`HH`) is good.

Danappelxx reviewed (changes requested):
Great PR! Just need to figure out the format and it's good to go.

JGiola commented:
No problem, for the mont and day I'm sure it can do if you use the enum specifiers, I don't know if it will do the same with the localised template...

JGiola committed:
Change format template

Change the template to include year, month, day and time zone and
removed unnecessary separators.

Now the time logged is something like:
10/07/16, 08:20:38.454 GMT+2 for en_US locale
07/10/16, 08:23:01,810 GMT+2 for it_IT locale

JGiola commented:
Done, I've even removed the separator from the template because they aren't needed, foundation is figuring out everything by itself.

And yes, the order of year, month and day are based on the locale 👍🏻

Danappelxx commented:
Awesome! Just one more thing - could you write a unit test that ensures the formatter spits out the right string? I believe what you say about the unnecessary separators but I'm worried that it might be different on Linux for whatever reason, etc.

JGiola commented:
The tests are no problem, but it possible to cheat a bit? 😁
Because the actual lotTimeFormatter is a fileprivate property I can't access directly inside the tests and in addition to that I have to set a fixed sets of locale and time zone to know how the strings would come out and so reset the dateFormat to it.

My idea is to tests on a brand new DateFormatter with a couple of different locale and time zone and two or three different dates to see if the format string without any separator is always equals to something we will know.
Is it ok for you?

Danappelxx commented:
If you use `@testable import` you should be able to access the `fileprivate` property. If this does not work, maybe make the formatter public? That could provide other benefits (user customization) anyway. 

I agree that you should use several constant dates with locales and just plug them into the formatter.

JGiola committed:
Change log date formatter access control

Move the logTimeFormatter property from fileprivate to public for
allowing customization

JGiola committed:
Add logTimeFormatter tests

Add tests for checking the formatter output against English and Italian
locales

JGiola commented:
`@testable import`allow only internal properties to be accessible. I've moved the `logTimeFormatter` property to be public so people can changes all the properties as they like.

The tests will check against the en_US and it_IT locales, I have to add another one?

JGiola committed:
Add missing Foundation import in LoggerTests

Danappelxx committed:
Clean up tests

JGiola committed:
Change hour format

Change HH to kk for forcing 24h cycle on Linux

JGiola commented:
I'm start to thinking that on Linux the unicode library is parsing the date skeleton template in a totally different way than on macOS.
It seems to completely ignore the forced 24 hour marker and for the Italian locale the two platform don't agree on the coma separator between date and hour...

JGiola commented:
Ok, I've spin up an ubuntu 12.04 server and I've found a super weird bug. On Linux if you add anyone of the time zone specifier to the skeleton format will disable the forced 24h format, as soon as you change the skeleton from `yyMMdd HHmmssSSSz` to `yyMMdd HHmmssSSS` the printed hours will revert correctly to the force 24h format.

Danappelxx commented:
Interesting... so what you're saying is that removing the z will fix it? Can you try committing that change? If it fixes CI, we can happily merge :)

JGiola committed:
Remove time zone specifier from format

Danappelxx commented:
Oh, so removing Z removes the time zone? That's a shame... @paulofaria @robertjpayne what do you think is the right thing to do here?

paulofaria commented:
@JGiola could you please raise a bug at bug.swift.org? I think the formatting itself is secondary in regards to correctness. Timezones should be respected above all. If this means different formatting, I'm fine with that.

JGiola commented:
Yeah, I've filed it when I've been able to check what was going on on a Ubuntu server.
If you want to join in and write it down better is here https://bugs.swift.org/browse/SR-2906

JGiola commented:
In regard of the pull request what you want me to do? Going back on adding the time zone to the skeleton format and comment out the tests for now? Or leave things as they are now?

paulofaria committed:
Merge branch 'master' into use-foundation-date-in-log-timestamp
